cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -191,38 +191,73 @@ VAR_LEN_TYPES(IS_NOT_NULL, isnotnull)
 #undef IS_NOT_NULL
 
 FORCE_INLINE
-const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
-                                    int64 offset64, int64 length, int32* out_len) {
-  if (length <= 0 || input == nullptr || in_len <= 0) {
-    *out_len = 0;
+const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_data_len,
+                                    int64 offset64, int64 length, int32* out_data_len) {
+  if (length <= 0 || input == nullptr || in_data_len <= 0) {
+    *out_data_len = 0;
+    return "";
+  }
+
+  int32 in_glyphs_count = utf8_length(context, input, in_data_len);
+
+  // handle invalid glyphs in input
+  if (in_glyphs_count == 0) {
+    *out_data_len = 0;
     return "";
   }
 
   int32 offset = static_cast<int32>(offset64);
-  int32 startIndex = offset - 1;  // offset is 1 for first char
+  int32 from_glyph = offset - 1;  // offset is 1 for first char
```

Comment:
There is a lot of overflow + wrap issues here. I'd recommend doing the whole computation in int64.

Reply:
Why is there a difference between offset and from_glyph here? The wrapping logic is not explained; I would expect

| offset64 | refers to |
|---|---|
| 0 | first glyph |
| 1 | second glyph |
| -1 | last glyph |

If I'm reading this right offset64 == 0 and offset64 == 1 currently have the same result, which surprises me. In any case, please add a comment explaining the semantics of offset64 carefully

Reply:
added comment. Thanks